src/display_context.cpp

```diff
@@ -87,21 +87,59 @@ display_context::can_move_result display_context::unit_can_move(const unit& u) c
 	const team& current_team = get_team(u.side());
 
 	can_move_result result = {false, false};
-	for(const map_location& adj : get_adjacent_tiles(u.get_location())) {
-		if (map().on_board(adj)) {
-			if(!result.attack_here) {
-				const unit_map::const_iterator i = units().find(adj);
-				if (i.valid() && !i->incapacitated() && current_team.is_enemy(i->side())) {
-					result.attack_here = true;
+
+	// Generating pairs of min-max ranges of all units weapons
+	const auto& attacks = u.attacks();
+	std::vector<std::pair<int, int>> distance_ranges;
+	for (const auto& attack : attacks) {
+		int min_range = attack.min_range();
+		int max_range = attack.max_range();
+		distance_ranges.emplace_back(min_range , max_range);
+	}
+
+	// Merging it here, may be removed if some checks for every weapon
+	// Should be done individually (Like line-of sight)
+	std::sort(distance_ranges.begin(), distance_ranges.end());
```

Comment:
What's the purpose of this sort?

Reply:
@hladkyjakub can you remove it and see if that changes anything?

Reply:
I think it's quite likely that it's required for the following merge to function. However, I think the merge can be eliminated too – see [my other comment](https://github.com/wesnoth/wesnoth/pull/9255#discussion_r1730632203).